### ISSUE TITLE
Add a Debug impl to everything

### ISSUE THREAD

**Issue by Lokathor** [gfx-rs/gfx#2560]
Every single type within `gfx-hal`, and all backends, should have a `Debug` impl available in the `0.2` release. This is not a controversial idea.

However, I would add that while we're setting it up we might want to make an effort to have them be _optional_ Debug impls, gated by a feature flag (probably one that is _on_ by default). Why? When there's a lot of types it ends up taking build time just to generate the Debug code. Even if the user doesn't call the Debug impl, it still gets generated and then thrown away. It's maybe not the _biggest_ deal, but it's worth consideration. This isn't totally crazy, `winapi` already does it.

Right now most types don't have Debug, and most types _might_ have content that isn't friendly to simply deriving Debug. That's fine, but we want them to have Debug anyway so that end users can derive a Debug even if they're holding a `gfx-hal` type in their struct. It's just the polite thing to do. What's needed is a "dummy" impl that just prints the type's name and a note that other info has been cut. This is a trivial macro_rules to write:

```rust
macro_rules! generate_dummy_Debug_for {
  ($t:ty) => {
    impl core::fmt::Debug for $t {
      /// This only prints the type name, with no field data
      fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
        write!(f, concat!(stringify!($t), " {{ fields omitted }} "))
      }
    }
  };
}
```

**Comment by seivan** [comment]
Is this the only thing missing from pushing master as a crate? 
https://github.com/gfx-rs/gfx/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22HAL+0.2+release%22

**Comment by kvark** [comment]
@seivan no, we just created the milestone, and it will surely be populated with more issues as we go. Release for hal-0.2 is going to be driven by the users: do you require it? if so, please explain (better in the chat, not here).

**Comment by kvark** [comment]
Considering fixed by #2459. Please feel free to reopen otherwise!
